### PR TITLE
Ensure foreground stats are recorded for handled errors

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Ensures that foreground stats are recorded correctly for handled errors. This fixes a bug where `app.inForeground`, `app.durationInForeground`, and `app.duration` were captured for unhandled errors but not for handled errors. This is due to the two errors using a different mechanism for capture.

## Changeset

- Added `[BSG_KSCrash captureApplicationStatsInfo]` to `BSG_KSSystemInfo` to ensure that application stats are always captured for handled errors
- Returned the value of `bsg_kscrw_i_writeAppStats` as a JSON string when capturing application stats
- Renamed `BSG_ThreadDataBuffer` to `BSG_JSONDataBuffer` as the struct can be used as a buffer for both threads + app stats

## Tests

Verified locally in an example app that the fields are now populated.

### CODE REVIEW

File: Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.h [hunk 1]
```diff
@@ -119,6 +119,11 @@
  */
 - (NSArray<BugsnagThread *> *)captureThreads:(NSException *)exc depth:(int)depth;
 
+/**
+ * Collects information about the application's foreground state (duration in foreground/background)
+ */
+- (NSDictionary *)captureApplicationStatsInfo;
```
Comment (tomlongridge) [inline Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.h:125]:
```suggestion
/**
 * Collects information about the application's state (e.g. duration in foreground/background)
 */
- (NSDictionary *)captureAppStats;
```

File: Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.h [hunk 2]
```diff
@@ -74,6 +74,13 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext);
  */
 char *bsg_kscrw_i_captureThreadTrace(const BSG_KSCrash_Context *crashContext);
 
+/**
+ * Captures app foreground stats for use by the Objective-C layer to append to handled errors.
+ *
+ * @return the stats encoded as a JSON string
+ */
+char *bsg_kscrw_i_captureAppForegroundStats(const BSG_KSCrash_Context *crashContext);
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.h:82]: Renamed all methods as specified

File: Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c [hunk 3]
```diff
@@ -1700,6 +1700,18 @@ char *bsg_kscrw_i_captureThreadTrace(const BSG_KSCrash_Context *crashContext) {
     return userData.data;
 }
 
+char *bsg_kscrw_i_captureAppForegroundStats(const BSG_KSCrash_Context *crashContext) {
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c]: I observed a warning about `bsg_kscrw_i_writeAppStats` not declaring its 3rd parameter as `const` so have addressed that - I'm assuming that's the same as what you observed.

**Comment by fractalwrench** [comment]
Updated approach to avoid serialization by getting the information from the KSCrash state directly. This necessitates a small refactor of the logic for updating the duration/durationInForeground fields, which need to be recalculated whenever an error is captured.

File: Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashState.m [hunk 4]
```diff
@@ -72,6 +72,8 @@
 // Avoiding static functions due to linker issues.
 
 // ============================================================================
+void bsg_kscrashstate_updateForegroundStats(BSG_KSCrash_State *const state);
```
Comment (tomlongridge) [inline Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashState.m:75]: Declaration can be removed.

File: Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashState.h [hunk 5]
```diff
@@ -121,6 +121,14 @@ void bsg_kscrashstate_notifyAppCrash(BSG_KSCrashType type);
  */
 const BSG_KSCrash_State *bsg_kscrashstate_currentState(void);
 
+/**
+ * Updates the stats for duration in foreground/background. This needs to
+ * be updated whenever an error report is captured.
+ *
+ * @param state the kscrash state
+ */
+void bsg_kscrashstate_updateForegroundStats(BSG_KSCrash_State *const state);
```
Comment (tomlongridge) [inline Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_KSCrashState.h:130]:
```suggestion
void bsg_kscrashstate_updateDurationStats(BSG_KSCrash_State *const state);
```

**Pull request merged**